NPE occurred when target Changed

When VTTablet Reparent

VtDriver will print following error message:
```
at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:427) ~[mybatis-spring-2.0.6.jar:2.0.6]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
at sun.reflect.GeneratedMethodAccessor205.invoke(Unknown Source) ~[?:?]
at org.apache.ibatis.session.defaults.DefaultSqlSession.insert(DefaultSqlSession.java:181) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:194) ~[mybatis-3.5.7.jar:3.5.7]
at com.sun.proxy.$Proxy207.update(Unknown Source) ~[?:?]
at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:64) ~[mybatis-3.5.7.jar:3.5.7]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
at sun.reflect.GeneratedMethodAccessor185.invoke(Unknown Source) ~[?:?]
at org.apache.ibatis.executor.CachingExecutor.update(CachingExecutor.java:76) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.BaseExecutor.update(BaseExecutor.java:117) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.SimpleExecutor.doUpdate(SimpleExecutor.java:50) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.statement.RoutingStatementHandler.update(RoutingStatementHandler.java:74) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.statement.PreparedStatementHandler.update(PreparedStatementHandler.java:47) ~[mybatis-3.5.7.jar:3.5.7]
at com.sun.proxy.$Proxy208.execute(Unknown Source) ~[?:?]
at org.apache.ibatis.logging.jdbc.PreparedStatementLogger.invoke(PreparedStatementLogger.java:59) ~[mybatis-3.5.7.jar:3.5.7]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
at sun.reflect.GeneratedMethodAccessor192.invoke(Unknown Source) ~[?:?]
at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java) ~[HikariCP-3.4.2.jar:?]
at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44) ~[HikariCP-3.4.2.jar:?]
at com.jd.jdbc.vitess.VitessPreparedStatement.execute(VitessPreparedStatement.java:413) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.vitess.VitessStatement.executeInternal(VitessStatement.java:433) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.execute(Executor.java:132) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.newExecute(Executor.java:484) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.insideTransaction(Executor.java:689) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.lambda$executePlan$0(Executor.java:340) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.engine.InsertEngine.execute(InsertEngine.java:212) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.engine.InsertEngine.execInsertSharded(InsertEngine.java:310) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.VcursorImpl.executeMultiShard(VcursorImpl.java:100) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.executeMultiShard(Executor.java:202) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.srvtopo.ScatterConn.executeMultiShard(ScatterConn.java:118) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.srvtopo.ScatterConn.multiGoTransaction(ScatterConn.java:325) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.srvtopo.ScatterConn.oneShard(ScatterConn.java:456) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.srvtopo.ScatterConn.lambda$executeMultiShard$0(ScatterConn.java:156) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.queryservice.RetryTabletQueryService.beginExecute(RetryTabletQueryService.java:112) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.queryservice.RetryTabletQueryService.retry(RetryTabletQueryService.java:304) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.queryservice.RetryTabletQueryService.lambda$beginExecute$2(RetryTabletQueryService.java:114) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.queryservice.CombinedQueryService.beginExecute(CombinedQueryService.java:162) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.queryservice.CombinedQueryService.getNativeQueryService(CombinedQueryService.java:82) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.queryservice.NativeQueryService.<init>(NativeQueryService.java:84) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.pool.StatefulConnectionPool.getStatefulConnectionPool(StatefulConnectionPool.java:72) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.pool.StatefulConnectionPool.<init>(StatefulConnectionPool.java:56) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.pool.HikariUtil.getHikariConfig(HikariUtil.java:51) ~[vtdriver-1.0.0.jar:?]
at com.jd.zaxxer.hikari.HikariConfig.setDataSourceProperties(HikariConfig.java:468) ~[HikariCP-1.0-20210311.110649-1.jar:?]
at java.util.Hashtable.putAll(Hashtable.java:522) ~[?:1.8.0_60]
Caused by: java.lang.NullPointerException
... 155 more
at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:427) ~[mybatis-spring-2.0.6.jar:2.0.6]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
at sun.reflect.GeneratedMethodAccessor205.invoke(Unknown Source) ~[?:?]
at org.apache.ibatis.session.defaults.DefaultSqlSession.insert(DefaultSqlSession.java:181) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:194) ~[mybatis-3.5.7.jar:3.5.7]
at com.sun.proxy.$Proxy207.update(Unknown Source) ~[?:?]
at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:64) ~[mybatis-3.5.7.jar:3.5.7]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
at sun.reflect.GeneratedMethodAccessor185.invoke(Unknown Source) ~[?:?]
at org.apache.ibatis.executor.CachingExecutor.update(CachingExecutor.java:76) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.BaseExecutor.update(BaseExecutor.java:117) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.SimpleExecutor.doUpdate(SimpleExecutor.java:50) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.statement.RoutingStatementHandler.update(RoutingStatementHandler.java:74) ~[mybatis-3.5.7.jar:3.5.7]
at org.apache.ibatis.executor.statement.PreparedStatementHandler.update(PreparedStatementHandler.java:47) ~[mybatis-3.5.7.jar:3.5.7]
at com.sun.proxy.$Proxy208.execute(Unknown Source) ~[?:?]
at org.apache.ibatis.logging.jdbc.PreparedStatementLogger.invoke(PreparedStatementLogger.java:59) ~[mybatis-3.5.7.jar:3.5.7]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
at sun.reflect.GeneratedMethodAccessor192.invoke(Unknown Source) ~[?:?]
at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java) ~[HikariCP-3.4.2.jar:?]
at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44) ~[HikariCP-3.4.2.jar:?]
at com.jd.jdbc.vitess.VitessPreparedStatement.execute(VitessPreparedStatement.java:413) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.vitess.VitessStatement.executeInternal(VitessStatement.java:433) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.execute(Executor.java:132) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.newExecute(Executor.java:484) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.insideTransaction(Executor.java:689) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.lambda$executePlan$0(Executor.java:340) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.engine.InsertEngine.execute(InsertEngine.java:212) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.engine.InsertEngine.execInsertSharded(InsertEngine.java:310) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.VcursorImpl.executeMultiShard(VcursorImpl.java:100) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.Executor.executeMultiShard(Executor.java:202) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.srvtopo.ScatterConn.executeMultiShard(ScatterConn.java:218) ~[vtdriver-1.0.0.jar:?]
at com.jd.jdbc.concurrency.AllErrorRecorder.throwException(AllErrorRecorder.java:86) ~[vtdriver-1.0.0.jar:?]
```

--- a/src/main/java/com/jd/jdbc/discovery/HealthCheck.java
+++ b/src/main/java/com/jd/jdbc/discovery/HealthCheck.java
@@ -328,6 +328,8 @@ public enum HealthCheck {
                 String oldTargetKey = keyFromTarget(preTarget);
                 this.healthData.get(oldTargetKey).remove(tabletAlias);
                 this.healthData.computeIfAbsent(targetKey, key -> new HashMap<>());
+                Topodata.Tablet tablet = th.getTablet().toBuilder().setType(th.getTarget().getTabletType()).setKeyspace(th.getTarget().getKeyspace()).setShard(th.getTarget().getShard()).build();
+                th.getQueryService().setTablet(tablet);
             }
             this.healthData.get(targetKey).put(tabletAlias, th);
             boolean isPrimary = th.getTarget().getTabletType() == Topodata.TabletType.MASTER;

--- a/src/main/java/com/jd/jdbc/queryservice/CombinedQueryService.java
+++ b/src/main/java/com/jd/jdbc/queryservice/CombinedQueryService.java
@@ -43,7 +43,7 @@ import java.util.Properties;
  * QueryService that combines the NativeProtocol and GRPC
  **/
 public class CombinedQueryService implements IQueryService, IHealthCheckQueryService {
-    private final Topodata.Tablet tablet;
+    private Topodata.Tablet tablet;
 
     private final IQueryService tabletQueryService;
 
@@ -51,14 +51,20 @@ public class CombinedQueryService implements IQueryService, IHealthCheckQuerySer
 
     public CombinedQueryService(ManagedChannel channel, Topodata.Tablet tablet) {
         this.tablet = tablet;
-        tabletQueryService = new TabletQueryService(channel);
-
+        this.tabletQueryService = new TabletQueryService(channel);
         SecurityCenter.Credential credential = SecurityCenter.INSTANCE.getCredential(this.tablet.getKeyspace());
-        Properties dsProperties = Config.getDataSourceConfig(tablet.getKeyspace(), credential.getUser(), tablet.getType());
-        if (dsProperties != null && Objects.equals(Topodata.TabletType.MASTER, tablet.getType()) && Objects.equals(tablet.getType(), VitessJdbcProperyUtil.getTabletType(dsProperties))) {
-            Properties properties = Config.getConnectionPoolConfig(tablet.getKeyspace(), credential.getUser(), tablet.getType());
+        Topodata.TabletType tabletType = tablet.getType();
+        Properties dsProperties = Config.getDataSourceConfig(tablet.getKeyspace(), credential.getUser(), tabletType);
+        if (dsProperties != null && Objects.equals(Topodata.TabletType.MASTER, tabletType) && Objects.equals(tabletType, VitessJdbcProperyUtil.getTabletType(dsProperties))) {
+            Properties properties = Config.getConnectionPoolConfig(tablet.getKeyspace(), credential.getUser(), tabletType);
             nativeQueryService = new NativeQueryService(tablet, credential.getUser(), credential.getPassword(), dsProperties, properties);
         }
+    }
+
+    @Override
+    public void setTablet(Topodata.Tablet tablet) {
+        this.tablet = tablet;
+        nativeQueryService.setTablet(tablet);
     }
 
     private IQueryService getNativeQueryService() {
@@ -71,7 +77,7 @@ public class CombinedQueryService implements IQueryService, IHealthCheckQuerySer
                 Topodata.TabletType tabletType = tablet.getType();
                 Properties properties = Config.getConnectionPoolConfig(tablet.getKeyspace(), credential.getUser(), tabletType);
                 Properties dsProperties = Config.getDataSourceConfig(tablet.getKeyspace(), credential.getUser(), tabletType);
-                if ((properties == null || dsProperties == null) && Objects.equals(Topodata.TabletType.RDONLY, tablet.getType())) {
+                if ((properties == null || dsProperties == null) && Objects.equals(Topodata.TabletType.RDONLY, tabletType)) {
                     properties = Config.getConnectionPoolConfig(tablet.getKeyspace(), credential.getUser(), Topodata.TabletType.REPLICA);
                     dsProperties = Config.getDataSourceConfig(tablet.getKeyspace(), credential.getUser(), Topodata.TabletType.REPLICA);
                 }

--- a/src/main/java/com/jd/jdbc/queryservice/IQueryService.java
+++ b/src/main/java/com/jd/jdbc/queryservice/IQueryService.java
@@ -26,6 +26,7 @@ import com.jd.jdbc.sqltypes.VtResultSet;
 import com.jd.jdbc.srvtopo.BindVariable;
 import com.jd.jdbc.srvtopo.BoundQuery;
 import io.vitess.proto.Query;
+import io.vitess.proto.Topodata;
 import java.sql.SQLException;
 import java.util.List;
 import java.util.Map;
@@ -57,4 +58,7 @@ public interface IQueryService extends IParentQueryService {
                                                 Query.ExecuteOptions options) throws Exception;
 
     Query.ReleaseResponse release(IContext context, Query.Target target, Long transactionID, Long reservedID) throws SQLException;
+
+    default void setTablet(Topodata.Tablet tablet) {
+    }
 }

--- a/src/main/java/com/jd/jdbc/queryservice/NativeQueryService.java
+++ b/src/main/java/com/jd/jdbc/queryservice/NativeQueryService.java
@@ -70,13 +70,18 @@ public class NativeQueryService implements IQueryService {
 
     private final StatefulConnectionPool statefulConnectionPool;
 
-    private final Topodata.Tablet tablet;
+    private Topodata.Tablet tablet;
 
     private Histogram.Timer histogramTimer = null;
 
     public NativeQueryService(final Topodata.Tablet tablet, final String user, final String password, final Properties dsProperties, final Properties properties) {
         this.tablet = tablet;
         this.statefulConnectionPool = StatefulConnectionPool.getStatefulConnectionPool(tablet, user, password, dsProperties, properties);
+    }
+
+    @Override
+    public void setTablet(Topodata.Tablet tablet) {
+        this.tablet = tablet;
     }
 
     //currently not in used

--- a/src/main/java/com/jd/jdbc/util/NetUtil.java
+++ b/src/main/java/com/jd/jdbc/util/NetUtil.java
@@ -1,6 +1,5 @@
 package com.jd.jdbc.util;
 
-import com.jd.jdbc.monitor.MonitorServer;
 import com.jd.jdbc.sqlparser.support.logging.Log;
 import com.jd.jdbc.sqlparser.support.logging.LogFactory;
 import java.io.BufferedReader;
@@ -18,15 +17,19 @@ import java.net.UnknownHostException;
 import java.util.Enumeration;
 
 public class NetUtil {
-    private static final Log log = LogFactory.getLog(MonitorServer.class);
+    private static final Log log = LogFactory.getLog(NetUtil.class);
 
-    public static String getLocalAdder() {
-        String localAddress = null;
+    private static String localAddress = null;
+
+    static {
         try {
             localAddress = getLocalHostAddress().getHostAddress();
-        } catch (Exception e) {
-            log.error(String.format("getLocalAddr error! causeby:%s", e.getMessage()));
+        } catch (UnknownHostException | SocketException e) {
+            log.error("getLocalAddr error! causeby:%s" + e.getMessage());
         }
+    }
+
+    public static String getLocalAdder() {
         return localAddress;
     }
 

--- a/src/main/java/com/jd/jdbc/vitess/VitessDriver.java
+++ b/src/main/java/com/jd/jdbc/vitess/VitessDriver.java
@@ -116,10 +116,7 @@ public class VitessDriver implements java.sql.Driver {
             SecurityCenter.INSTANCE.addCredential(prop);
             String defaultKeyspace = keySpaces.get(0);
 
-            String role = prop.getProperty(DRIVER_PROPERTY_ROLE_KEY, DRIVER_PROPERTY_ROLE_RW);
-            if (!prop.containsKey(DRIVER_PROPERTY_ROLE_KEY)) {
-                prop.put(DRIVER_PROPERTY_ROLE_KEY, role);
-            }
+            String role = VitessJdbcProperyUtil.getRole(prop);
             Topodata.TabletType tabletType = VitessJdbcProperyUtil.getTabletType(prop);
             Config.setConfig(prop, defaultKeyspace, SecurityCenter.INSTANCE.getCredential(defaultKeyspace).getUser(), tabletType);
             TopoServer topoServer = Topo.getTopoServer(Topo.TopoServerImplementType.TOPO_IMPLEMENTATION_ETCD2, "http://" + prop.getProperty("host") + ":" + prop.getProperty("port"));

--- a/src/main/java/com/jd/jdbc/vitess/VitessJdbcProperyUtil.java
+++ b/src/main/java/com/jd/jdbc/vitess/VitessJdbcProperyUtil.java
@@ -98,7 +98,7 @@ public class VitessJdbcProperyUtil {
     }
 
     public static Topodata.TabletType getTabletType(Properties props) {
-        String role = props.getProperty(Constant.DRIVER_PROPERTY_ROLE_KEY, Constant.DRIVER_PROPERTY_ROLE_RW);
+        String role = getRole(props);
         switch (role.toLowerCase()) {
             case Constant.DRIVER_PROPERTY_ROLE_RW:
                 return Topodata.TabletType.MASTER;

--- a/src/main/java/com/jd/jdbc/vitess/VitessPreparedStatement.java
+++ b/src/main/java/com/jd/jdbc/vitess/VitessPreparedStatement.java
@@ -57,8 +57,6 @@ import java.util.stream.Collectors;
 import lombok.AllArgsConstructor;
 import lombok.Data;
 
-import static com.jd.jdbc.common.Constant.DRIVER_PROPERTY_ROLE_KEY;
-
 public class VitessPreparedStatement extends AbstractVitessPreparedStatement {
     private static final Log LOGGER = LogFactory.getLog(VitessPreparedStatement.class);
 
@@ -88,8 +86,7 @@ public class VitessPreparedStatement extends AbstractVitessPreparedStatement {
         this.sqls = split(inputSQL);
 
         for (int i = 0; i < sqls.size(); i++) {
-            try (InnerConnection innerConnection = StatefulConnectionPool.getJdbcConnection(this.connection.getDefaultKeyspace(),
-                (Topodata.TabletType) context.getContextValue(DRIVER_PROPERTY_ROLE_KEY))) {
+            try (InnerConnection innerConnection = StatefulConnectionPool.getJdbcConnection(this.connection.getDefaultKeyspace(), RoleUtils.getTabletType(context))) {
                 JdbcConnection connectionImpl = innerConnection.getConnectionImpl();
                 String encoding = this.connection.getProperties().getProperty(VitessPropertyKey.CHARACTER_ENCODING.getKeyName());
                 Session session = connectionImpl.getSession();

